CDP4Composition/Mvvm/ViewModelBase.cs

```diff
@@ -95,22 +112,111 @@ protected ViewModelBase()
+        }
+
+        /// <summary>
+        /// Initializes this instance of <see cref="ViewModelBase{T}"/>
+        /// </summary>
+        /// <param name="thing">The <see cref="Thing"/></param>
+        /// <param name="session">The <see cref="ISession"/></param>
+        private void Initialize(T thing, ISession session)
+        {
+            this.BeforeInitialize();
             this.ThemeName = AppliedTheme.ThemeName;
-            
+
             this.PermissionService = session.PermissionService;
             this.Disposables = new List<IDisposable>();
-            this.Thing = thing;            
+            this.Thing = thing;
             this.Session = session;
 
             this.RevisionNumber = thing.RevisionNumber;
             this.IDalUri = thing.IDalUri;
 
-            var thingSubscription = CDPMessageBus.Current.Listen<ObjectChangedEvent>(this.Thing)
-                .Where(objectChange => objectChange.EventKind == EventKind.Updated && objectChange.ChangedThing.RevisionNumber > this.RevisionNumber)
-                .ObserveOn(RxApp.MainThreadScheduler)
-                .Subscribe(this.ObjectChangeEventHandler);
-            this.Disposables.Add(thingSubscription);
+            Func<ObjectChangedEvent, bool> discriminator =
+                objectChange =>
+                objectChange.EventKind == EventKind.Updated 
+                && objectChange.ChangedThing.RevisionNumber > this.RevisionNumber;
+
+            Action<ObjectChangedEvent> action = this.ObjectChangeEventHandler;
+
+            if (this.AllowMessageBusSubscriptions)
+            {
+                var thingSubscription = CDPMessageBus.Current.Listen<ObjectChangedEvent>(this.Thing)
+                    .Where(discriminator)
+                    .ObserveOn(RxApp.MainThreadScheduler)
+                    .Subscribe(action);
+
+                this.Disposables.Add(thingSubscription);
+            }
+            else
+            {
+                var thingObserver = CDPMessageBus.Current.Listen<ObjectChangedEvent>(typeof(T));
+
+                this.Disposables.Add(this.MessageBusHandler.GetHandler<ObjectChangedEvent>().RegisterEventHandler(
+                    thingObserver, 
+                    new ObjectChangedMessageBusHandlerData(
+                        this.Thing, 
+                        discriminator,
+                        action)));
+            }
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="logger"/>
+        /// </summary>
+        private void InitializeLogger()
+        {
+            logger = LogManager.GetLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Execute code that needs te run before all code in the hierarchy of constructors is executed, but after a ContainerViewModel was added
+        /// </summary>
+        protected virtual void BeforeInitialize()
+        {
+            var searchContainerViewModel = this.ContainerViewModel;
```

Comment:
not clear to me why this is called `searchContainerViewModel`, what is the relationship with search?

Reply:
Idea was to search up the ViewModel tree hierarchy for an object that implements IHaveMessageBusHandler, so that could be used by the child RowViewModels. That's the reason the the term "search" was added. But if it confuses it more, that's not good, so I removed it.